### ISSUE TITLE
Generic solver infers `Unknown` only when using new-style generics

### ISSUE THREAD

**Issue by wence-** [astral-sh/ty#1353]
### Summary

I think this is not reported previously.

Consider (https://play.ty.dev/79ca09f4-57e4-4544-abda-3ef3be0c08bd):

```python
from __future__ import annotations

from typing import Self, reveal_type, Protocol, TypeVar, Generic


class PayloadImplicit(Protocol):
    @classmethod
    def from_message(cls, message: MessageImplicit[Self]) -> Self:
        ...
    
class MessageImplicit[T: PayloadImplicit]:
    pass

class Implicit:
    @classmethod
    def from_message(cls, message: MessageImplicit[Self]) -> Self:
        return cls()


reveal_type(MessageImplicit[Implicit]())
reveal_type(Implicit.from_message(MessageImplicit[Implicit]()))


class PayloadExplicit(Protocol):
    @classmethod
    def from_message(cls, message: MessageExplicit[Self]) -> Self:
        ...

T = TypeVar("T", bound="PayloadExplicit")

class MessageExplicit(Generic[T]):
    pass

class Explicit:
    @classmethod
    def from_message(cls, message: MessageExplicit[Self]) -> Self:
        return cls()

reveal_type(MessageExplicit[Explicit]())
reveal_type(Explicit.from_message(MessageExplicit[Explicit]()))
```

To the best of my understanding, the definitions of `MessageImplicit` and `MessageExplicit` are structurally identical. However, `ty` (correctly) reveals that the type of `Explicit.from_message(...)` is `Explicit` whereas reveals `Implicit.from_message(...)` as `Unknown`

### Version

9090aead0

**Comment by carljm** [comment]
The [same discrepancy appears](https://play.ty.dev/51ed20b7-6546-4d4e-8b5d-e819bb6669b7) without the upper bounds on the typevars (and thus without the protocols):

```py
from __future__ import annotations

from typing import Self, reveal_type, TypeVar, Generic


class MessageImplicit[T]:
    pass

class Implicit:
    @classmethod
    def from_message(cls, message: MessageImplicit[Self]) -> Self:
        return cls()

reveal_type(MessageImplicit[Implicit]())
reveal_type(Implicit.from_message(MessageImplicit[Implicit]()))

T = TypeVar("T")

class MessageExplicit(Generic[T]):
    pass

class Explicit:
    @classmethod
    def from_message(cls, message: MessageExplicit[Self]) -> Self:
        return cls()

reveal_type(MessageExplicit[Explicit]())
reveal_type(Explicit.from_message(MessageExplicit[Explicit]()))
```

So it looks like just a problem in typevar solving. We will probably want to look at it again if it still repros after #623

**Comment by carljm** [comment]
Thanks for the report!

**Comment by carljm** [comment]
This is fixed.
